pkg/volume/portworx/portworx_util.go

```diff
@@ -0,0 +1,276 @@
+		default:
+			labels[strings.ToLower(k)] = v
+		}
+	}
+	source := osdapi.Source{}
+	locator := osdapi.VolumeLocator{
+		Name:         p.options.PVName,
+		VolumeLabels: labels,
+	}
+	volumeID, err := client.Create(&locator, &source, &spec)
+	if err != nil {
+		glog.Infof("Error in Volume Create : %v", err)
+	}
+	return volumeID, requestGB, nil, err
+}
+
+func (util *PortworxVolumeUtil) osdClient(hostName string) (osdvolume.VolumeDriver, error) {
+	if util.portworxClient == nil {
+		var clientUrl string
+		if !strings.HasPrefix(hostName, "http://") {
+			clientUrl = "http://" + hostName + ":" + osdMgmtPort
+		} else {
+			clientUrl = hostName + ":" + osdMgmtPort
+		}
+		
+		driverClient, err := volumeclient.NewDriverClient(clientUrl, pxdDriverName, osdDriverVersion)
+		if err != nil {
+			return nil, err
+		}
+		util.portworxClient = driverClient
+	}
+
+	return volumeclient.VolumeDriver(util.portworxClient), nil
+}
+
+func (util *PortworxVolumeUtil) AttachVolume(m *portworxVolumeMounter) (string, error) {
+	hostName := m.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return "", err
+	}
+
+	devicePath, err := client.Attach(m.volumeID)
+	if err != nil {
+		if err == osdvolume.ErrVolAttachedOnRemoteNode {
+			// Volume is already attached to node.
+			glog.V(2).Infof("Attach operation is unsuccessful. Volume %q is already attached to another node.", m.volumeID)
+			return "", err
+		}
+		glog.V(2).Infof("AttachVolume on %v failed with error %v", m.volumeID, err)
+		return "", err
+	}
+	return devicePath, nil
+}
+
+func (util *PortworxVolumeUtil) DetachVolume(u *portworxVolumeUnmounter, deviceName string) error {
+	hostName := u.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	volumeID, err := getVolumeIDFromDeviceName(deviceName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Detach(volumeID)
+	if err != nil {
+		glog.V(2).Infof("DetachVolume on %v failed with error %v", u.volumeID, err)
+		return err
+	}
+	return nil
+}
+
+func (util *PortworxVolumeUtil) MountVolume(m *portworxVolumeMounter, mountPath string) error {
+	hostName := m.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Mount(m.volumeID, mountPath)
+	if err != nil {
+		glog.V(2).Infof("MountVolume on %v failed with error %v", m.volumeID, err)
+		return err
+	}
+	return nil
+}
+
+func (util *PortworxVolumeUtil) UnmountVolume(u *portworxVolumeUnmounter, deviceName, mountPath string) error {
+	hostName := u.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return err
+	}
+
+	volumeID, err := getVolumeIDFromDeviceName(deviceName)
+	if err != nil {
+		return err
+	}
+
+	err = client.Unmount(volumeID, mountPath)
+	if err != nil {
+		glog.V(2).Infof("UnmountVolume on mountPath: %v, VolName: %v failed with error %v", mountPath, u.volName, err)
+		return err
+	}
+	return nil
```

Comment:
K8s calls detach and unmount as a part of TearDown() function in portworx.go
will take a look at docker.go